src/libstd/path.rs

```diff
@@ -0,0 +1,2579 @@
+                }
+                State::Body => {
+                    self.back = State::Root;
+                }
+                State::Root => {
+                    self.back = State::Prefix;
+                    if self.has_physical_root {
+                        self.path = &self.path[.. self.path.len() - 1];
+                        return Some(Component::RootDir)
+                    } else if let Some(p) = self.prefix {
+                        if p.has_implicit_root() && !p.is_verbatim() {
+                            return Some(Component::RootDir)
+                        }
+                    }
+                }
+                State::Prefix if self.prefix_len() > 0 => {
+                    self.back = State::Done;
+                    return Some(Component::Prefix(unsafe {
+                        u8_slice_as_os_str(self.path)
+                    }))
+                }
+                State::Prefix => {
+                    self.back = State::Done;
+                    return None
+                }
+                State::Done => unreachable!()
+            }
+        }
+        None
+    }
+}
+
+fn optional_path(path: &Path) -> Option<&Path> {
+    if path.as_u8_slice().is_empty() { None } else { Some(path) }
+}
+
+impl<'a> cmp::PartialEq for Components<'a> {
+    fn eq(&self, other: &Components<'a>) -> bool {
+        iter::order::eq(self.clone(), other.clone())
+    }
+}
+
+impl<'a> cmp::Eq for Components<'a> {}
+
+impl<'a> cmp::PartialOrd for Components<'a> {
+    fn partial_cmp(&self, other: &Components<'a>) -> Option<cmp::Ordering> {
+        iter::order::partial_cmp(self.clone(), other.clone())
+    }
+}
+
+impl<'a> cmp::Ord for Components<'a> {
+    fn cmp(&self, other: &Components<'a>) -> cmp::Ordering {
+        iter::order::cmp(self.clone(), other.clone())
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Basic types and traits
+////////////////////////////////////////////////////////////////////////////////
+
+/// An owned, mutable path (akin to `String`).
+///
+/// This type provides methods like `push` and `set_extension` that mutate the
+/// path in place. It also implements `Deref` to `Path`, meaning that all
+/// methods on `Path` slices are available on `PathBuf` values as well.
+///
+/// More details about the overall approach can be found in
+/// the module documentation.
+///
+/// # Example
+///
+/// ```rust
+/// use std::path::PathBuf;
+///
+/// let mut path = PathBuf::new();
+/// path.push("c:\\");
+/// path.push("windows");
+/// path.push("system32");
+/// path.set_extension("dll");
+/// ```
+#[derive(Clone, Hash)]
+pub struct PathBuf {
+    inner: OsString
+}
+
+impl PathBuf {
+    fn as_mut_vec(&mut self) -> &mut Vec<u8> {
+        unsafe { mem::transmute(self) }
+    }
+
+    /// Allocate a `PathBuf` with initial contents given by the
+    /// argument.
+    pub fn new<S: ?Sized + AsOsStr>(s: &S) -> PathBuf {
```

Comment:
From our discussion the other day, this would be an interesting point to take `IntoOsString` vs `AsOsString` (a possible future modification).